### PR TITLE
GT-349: Improve the UX by moving the tract pages before making the new ones

### PR DESCRIPTION
**Pull request by pablomarti**

### CODE REVIEW

**Comment by pablomarti** [comment]
Hey @ryancarlson please test it and let me know what do you think with this behaviour, thanks!

**Comment by pablomarti** [comment]
Hey @ryancarlson it is still not perfect, but I believe is a lot better, would you please test it and let me know if I should continue getting this better or leave it like this? thanks!

File: godtools/ViewControllers/Tract/TractViewControllerGestures.swift [hunk 1]
```diff
@@ -23,6 +23,7 @@ extension TractViewController {
     
     @objc private func handleGesture(sender: UISwipeGestureRecognizer) {
         if sender.direction == .right {
+            print("New gesture")
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerGestures.swift]: Can you remove this debug please, or wrap it?

File: godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift [hunk 2]
```diff
@@ -21,28 +22,44 @@ extension TractViewController {
...
         
         self.currentPage = page
-        reloadPagesViews()
-        moveViews()
+        _ = reloadPagesViews()
+        _ = moveViews()
     }
     
     func moveToNextPage() {
+        print("Move next")
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift]: Can you remove this debug please, or wrap it?

File: godtools/Views/TractElements/TractCallToAction+Actions.swift [hunk 3]
```diff
@@ -11,6 +11,7 @@ import Foundation
 extension TractCallToAction {
     
     func moveToNextView() {
+        print("Move to next view")
```
Comment (ryancarlson) [inline godtools/Views/TractElements/TractCallToAction+Actions.swift]: Can you remove this debug please, or wrap it?

File: godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift [hunk 4]
```diff
@@ -21,28 +22,44 @@ extension TractViewController {
...
         if self.currentPage >= totalPages() - 1 {
+            print("Fail and return")
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift]: Can you remove this debug please, or wrap it?

File: godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift [hunk 5]
```diff
@@ -21,28 +22,44 @@ extension TractViewController {
...
             return
         }
         
-        self.currentPage += 1
-        reloadPagesViews()
-        moveViews()
+        _ = self.moveForewards()
+            .then { (success) -> Promise<Bool> in
+                if success == true {
+                    print("Reload pages")
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift]: Can you remove this debug please, or wrap it?

File: godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift [hunk 6]
```diff
@@ -21,28 +22,44 @@ extension TractViewController {
...
+            print("Fail and return")
             return
         }
         
-        self.currentPage += 1
-        reloadPagesViews()
-        moveViews()
+        _ = self.moveForewards()
+            .then { (success) -> Promise<Bool> in
+                if success == true {
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift]: why not `if success {` ? :)

File: godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift [hunk 7]
```diff
@@ -21,28 +22,44 @@ extension TractViewController {
...
+                    print("Reload pages")
+                    _ = self.reloadPagesViews()
+                    return Promise(value: true)
+                }
+                print("Fail Reload pages")
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift]: Can you remove this debug please, or wrap it?

File: godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift [hunk 8]
```diff
@@ -21,28 +22,44 @@ extension TractViewController {
...
             return
         }
         
-        self.currentPage -= 1
-        reloadPagesViews()
-        moveViews()
+        _ = self.moveBackwards()
+            .then { (success) -> Promise<Bool> in
+                if success == true {
+                    print("Reload pages")
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift]: Can you remove this debug please, or wrap it?

File: godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift [hunk 9]
```diff
@@ -21,28 +22,44 @@ extension TractViewController {
...
+                    _ = self.reloadPagesViews()
+                    return Promise(value: true)
+                }
+                print("Fail Reload pages")
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift]: Can you remove this debug please, or wrap it?

File: godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift [hunk 10]
```diff
@@ -60,28 +77,79 @@ extension TractViewController {
...
+        }).then { _ in
+            self.moveViewsExceptCurrentViews(pageViews: [currentPageView])
+            return Promise(value: true)
+        }
     }
     
-    func notifyCurrentViewDidAppearOnTract() {
-        let currentTractView = self.view.viewWithTag(self.currentPage + self.viewTagOrigin) as! TractView
-        currentTractView.contentView?.notifyViewDidAppearOnTract()
+    fileprivate func moveBackwards() -> Promise<Bool> {
```
Comment (ryancarlson) [inline godtools/ViewControllers/Tract/TractViewControllerPageMovements.swift]: This method and `moveForewards()` seem very similar. is there any way to reduce some of the duplicated logic?

**Comment by pablomarti** [comment]
Hey @ryancarlson I just uploaded the changes :D

**Pull request merged**